File: C#/Market.cs
using System;
using System.Collections.Generic;

namespace Assignment_C_
{
    class Market
    {
        public readonly Dictionary<int, Account> activeaccounts = new Dictionary<int, Account>();
        public readonly Dictionary<int, Account> allaccounts = new Dictionary<int, Account>();

        //Create account DONE
        public void CreateAccount(int accountNumber)
        {

            lock(activeaccounts){
                if (findStock()==-1){
                    Account account = new Account(accountNumber,true);
                    activeaccounts.Add(accountNumber, account);
                    allaccounts.Add(accountNumber, account);
                }else{
                    Account account = new Account(accountNumber,false);
                    activeaccounts.Add(accountNumber, account);
                    allaccounts.Add(accountNumber, account);
                }
            }
            

            
        }
        //Account list NEED FIX
        public int[] GetListOfAccounts()
        {
            lock(activeaccounts){

            List<int> result = new List<int>();
            foreach (Account account in activeaccounts.Values)
                result.Add(account.AccountNumber);

            return result.ToArray();

            }
        }






        //Find stock DONE
        public int findStock()
        {
            lock (activeaccounts) {
                foreach (Account account in activeaccounts.Values){
                    if (account.Stock){
                        return account.AccountNumber;
                    }

                }
                    
                return -1;
        }
    }



        //Leave market DONE
        public void LeaveMarket(int ID){
            lock(activeaccounts){
                activeaccounts.Remove(ID);
                
            }
        }



        //Give stock DONE
        public void GiveRandomStock(){
            lock(activeaccounts){
                if(activeaccounts.Count>0){
                    Random rand = new Random();
                    int randomint =  rand.Next(activeaccounts.Count);
                    int count = 0;
                    foreach (Account account in activeaccounts.Values){
                        if(count==randomint){
                            account.Stock=true;
                        }
                        count++;
                    }
                }
            }
        }



        //Transfer  DONE
        public void Transfer(int fromAccount, int toAccount)
        {
            lock (activeaccounts)
            {
                if (activeaccounts[fromAccount].Stock == false)
                    throw new Exception($"You do not have the stock");
                else if (!activeaccounts.ContainsKey(toAccount))
                    throw new Exception(
                        $"That account does not exist");
                else{
                    activeaccounts[fromAccount].Stock=false;
                    activeaccounts[toAccount].Stock=true;
                }
            }
        }


    }
}

File: Java/Server/Market.java
package Assignment;


import java.util.Random;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

public class Market {

    public static final Map<Integer, Account> activeaccounts = new TreeMap<>();
    public static final Map<Integer, Account> allaccounts = new TreeMap<>();

    //Creates an account and adds them to the account maps
    public static void createAccount(int accountNumber)
    {
        synchronized (activeaccounts) {
            if (Market.findStock() == -1) {
                Account account = new Account(accountNumber, true);
                activeaccounts.put(accountNumber, account);
                allaccounts.put(accountNumber, account);
            } else {
                Account account = new Account(accountNumber, false);
                activeaccounts.put(accountNumber, account);
                allaccounts.put(accountNumber, account);
            }
        }

    }
    //Removes an account from the active accounts map
    public void leaveMarket(int customerId){
        synchronized (activeaccounts) {
            activeaccounts.remove(customerId);
        }
    }

    //Give Stock to a random user
    public void giveRandomStock(){
        synchronized (activeaccounts) {
            if (activeaccounts.size()>0) {
                Random rand = new Random();
                int randomint = rand.nextInt(activeaccounts.size());
                int count = 0;
                for (Account account : activeaccounts.values()) {
                    if (count == randomint) {
                        account.setStock(true);
                    }
                    count++;
                }
            }
        }
    }

    public List<Integer> getListOfAccounts() {
        synchronized (activeaccounts) {
            List<Integer> result = new ArrayList<Integer>();

            for (Account account : activeaccounts.values())
                result.add(account.getAccountNumber());

            return result;
        }
    }
    public static int findStock(){
        synchronized (activeaccounts) {
            for (Account account : activeaccounts.values())
                if (account.getStock()) {
                    return account.getAccountNumber();
                }
            return -1;
        }
    }
    public  int findStock2(){
        synchronized (activeaccounts) {
            for (Account account : activeaccounts.values())
                if (account.getStock()) {
                    return account.getAccountNumber();
                }
            return -1;
        }
    }








    public void transfer(int fromAccount, int toAccount) throws Exception {
        synchronized (activeaccounts)
        {
            //Check account has the stock
            if (!activeaccounts.get(fromAccount).getStock()) {
                throw new Exception("You do not have the stock");
            }
            //Checking account transfering to exists
            else if (!activeaccounts.containsKey(toAccount)){
                throw new Exception("That account does not exist");
            }
            else {
                activeaccounts.get(fromAccount).setStock(false);
                activeaccounts.get(toAccount).setStock(true);
            }

        }
    }



}
